File: src/core/source_ledger.py
"""不可变源载荷与 canonical source 清单。"""

from __future__ import annotations

import hashlib
import json
import re
from pathlib import Path
from typing import Iterable, Mapping

from .preprocessor import SourceDocument


SOURCE_LEDGER_SCHEMA = "v5-source-ledger-1"
SUPPORTED_SOURCE_LANGUAGES = frozenset({"en", "fr", "de", "es", "ru", "ja", "und"})
ALLOWED_EXCLUDED_RAW_POLICIES = frozenset(
    {
        "UTF8_BOM",
        "UTF16_LE_BOM",
        "UTF16_BE_BOM",
        "UTF32_LE_BOM",
        "UTF32_BE_BOM",
        "EPUB_NON_SPINE_DATA",
        "DOCX_NON_DOCUMENT_DATA",
    }
)


def _sha256(payload: bytes) -> str:
    return hashlib.sha256(payload).hexdigest()


def normalize_source_language(value: str) -> str:
    """Return the supported primary BCP-47 source-language subtag."""
    if not isinstance(value, str) or not re.fullmatch(
        r"[A-Za-z]{2,3}(?:-[A-Za-z0-9]{2,8})*", value
    ):
        raise ValueError(f"invalid source language: {value!r}")
    primary = value.split("-", 1)[0].lower()
    if primary not in SUPPORTED_SOURCE_LANGUAGES:
        raise ValueError(f"unsupported source language: {value!r}")
    return primary


def _validate_canonical_segments(
    segments: Iterable[Mapping[str, object]], canonical_chars: int
) -> list[dict]:
    checked: list[dict] = []
    cursor = 0
    for index, original in enumerate(segments):
        segment = dict(original)
        start = segment.get("canonical_start")
        end = segment.get("canonical_end")
        if not isinstance(start, int) or isinstance(start, bool):
            raise ValueError(f"canonical segment {index} has invalid start")
        if not isinstance(end, int) or isinstance(end, bool):
            raise ValueError(f"canonical segment {index} has invalid end")
        if start != cursor:
            raise ValueError(
                f"canonical segments must be contiguous: expected {cursor}, got {start}"
            )
        if end < start or end > canonical_chars:
            raise ValueError(f"canonical segment {index} is out of bounds")
        for field in ("origin_kind", "origin_ref", "transformation"):
            if not isinstance(segment.get(field), str) or not segment[field]:
                raise ValueError(f"canonical segment {index} has invalid {field}")
        checked.append(segment)
        cursor = end

    if cursor != canonical_chars:
        raise ValueError(
            f"canonical segments must cover [0, {canonical_chars}), stopped at {cursor}"
        )
    return checked


def _validate_excluded_raw_ranges(
    ranges: Iterable[Mapping[str, object]], raw_size: int
) -> list[dict]:
    checked: list[dict] = []
    previous_end = 0
    for index, original in enumerate(ranges):
        item = dict(original)
        start = item.get("raw_start")
        end = item.get("raw_end")
        policy = item.get("policy")
        if policy not in ALLOWED_EXCLUDED_RAW_POLICIES:
            raise ValueError(f"unknown excluded raw range policy: {policy!r}")
        if not isinstance(start, int) or isinstance(start, bool):
            raise ValueError(f"excluded raw range {index} has invalid start")
        if not isinstance(end, int) or isinstance(end, bool):
            raise ValueError(f"excluded raw range {index} has invalid end")
        if start < previous_end or end <= start or end > raw_size:
            raise ValueError(f"excluded raw range {index} is overlapping or out of bounds")
        checked.append(item)
        previous_end = end
    return checked


def _write_immutable(path: Path, payload: bytes) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("xb") as target:
        target.write(payload)


def create_source_ledger(
    source: Path,
    project_dir: Path,
    document: SourceDocument,
    source_language: str = "en",
) -> dict:
    """保存原始字节、canonical source 和经验证的 provenance 清单。"""
    source = Path(source)
    project_dir = Path(project_dir)
    raw = source.read_bytes()
    if (
        document.raw_size is not None
        and document.raw_size != len(raw)
    ) or (
        document.raw_sha256 is not None
        and document.raw_sha256 != _sha256(raw)
    ):
        raise ValueError("source changed after document load")
    canonical = document.text.replace("\r\n", "\n").replace("\r", "\n")
    canonical_bytes = canonical.encode("utf-8")
    source_language = normalize_source_language(source_language)

    segments = _validate_canonical_segments(
        document.canonical_segments, len(canonical)
    )
    excluded_ranges = _validate_excluded_raw_ranges(
        document.excluded_raw_ranges, len(raw)
    )

    raw_target = project_dir / "source" / f"original{source.suffix.lower()}"
    canonical_target = project_dir / "source.txt"
    manifest_target = project_dir / "source_manifest.json"
    raw_relative = raw_target.relative_to(project_dir).as_posix()

    manifest = {
        "schema_version": SOURCE_LEDGER_SCHEMA,
        "coordinate_unit": "unicode_scalar",
        "raw_path": raw_relative,
        "raw_size": len(raw),
        "raw_sha256": _sha256(raw),
        "source_format": document.source_format,
        "encoding": document.encoding,
        "extractor": document.extractor,
        "sourceLanguage": source_language,
        "canonical_path": "source.txt",
        "canonical_chars": len(canonical),
        "canonical_sha256": _sha256(canonical_bytes),
        "canonical_segments": segments,
        "excluded_raw_ranges": excluded_ranges,
    }
    manifest_bytes = json.dumps(
        manifest, ensure_ascii=False, indent=2
    ).encode("utf-8")

    _write_immutable(raw_target, raw)
    _write_immutable(canonical_target, canonical_bytes)
    _write_immutable(manifest_target, manifest_bytes)
    return manifest

File: translator-v5/src/source/source-ledger.ts
import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";
import { dirname, isAbsolute, relative, resolve } from "node:path";
import { TextDecoder } from "node:util";

import type {
  CanonicalSegment,
  ExcludedRawRange,
  ScalarSource,
} from "./types.js";
import { UnicodeScalarMap } from "./types.js";

const SCHEMA_VERSION = "v5-source-ledger-1";
const COORDINATE_UNIT = "unicode_scalar";

export const ALLOWED_EXCLUDED_RAW_POLICIES: ReadonlySet<string> = new Set([
  "UTF8_BOM",
  "UTF16_LE_BOM",
  "UTF16_BE_BOM",
  "UTF32_LE_BOM",
  "UTF32_BE_BOM",
  "EPUB_NON_SPINE_DATA",
  "DOCX_NON_DOCUMENT_DATA",
]);

export type SourceIntegrityCode =
  | "MANIFEST_INVALID"
  | "MANIFEST_SCHEMA_UNSUPPORTED"
  | "COORDINATE_UNIT_UNSUPPORTED"
  | "MANIFEST_PATH_INVALID"
  | "RAW_SIZE_MISMATCH"
  | "RAW_HASH_MISMATCH"
  | "CANONICAL_HASH_MISMATCH"
  | "CANONICAL_UTF8_INVALID"
  | "CANONICAL_CHAR_COUNT_MISMATCH"
  | "CANONICAL_SEGMENTS_INVALID"
  | "EXCLUDED_POLICY_UNKNOWN"
  | "EXCLUDED_RANGE_INVALID";

export class SourceIntegrityError extends Error {
  readonly code: SourceIntegrityCode;

  constructor(code: SourceIntegrityCode, message: string) {
    super(`${code}: ${message}`);
    this.name = "SourceIntegrityError";
    this.code = code;
  }
}

function sha256(payload: string | Uint8Array): string {
  return createHash("sha256").update(payload).digest("hex");
}

function record(value: unknown, context: string): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new SourceIntegrityError("MANIFEST_INVALID", `${context} must be an object`);
  }
  return value as Record<string, unknown>;
}

function stringField(
  manifest: Record<string, unknown>,
  name: string,
): string {
  const value = manifest[name];
  if (typeof value !== "string" || value.length === 0) {
    throw new SourceIntegrityError("MANIFEST_INVALID", `${name} must be a non-empty string`);
  }
  return value;
}

function integerField(
  input: Record<string, unknown>,
  name: string,
): number {
  const value = input[name];
  if (!Number.isSafeInteger(value)) {
    throw new SourceIntegrityError("MANIFEST_INVALID", `${name} must be a safe integer`);
  }
  return value as number;
}

function manifestMemberPath(
  manifestDirectory: string,
  member: string,
): string {
  const target = resolve(manifestDirectory, member);
  const child = relative(manifestDirectory, target);
  if (child === "" || child.startsWith("..") || isAbsolute(child)) {
    throw new SourceIntegrityError(
      "MANIFEST_PATH_INVALID",
      `manifest member escapes its directory: ${member}`,
    );
  }
  return target;
}

function validateSegments(
  input: unknown,
  canonicalChars: number,
  rawSize: number,
): CanonicalSegment[] {
  if (!Array.isArray(input)) {
    throw new SourceIntegrityError(
      "CANONICAL_SEGMENTS_INVALID",
      "canonical_segments must be an array",
    );
  }
  const segments: CanonicalSegment[] = [];
  let cursor = 0;
  for (let index = 0; index < input.length; index += 1) {
    let item: Record<string, unknown>;
    try {
      item = record(input[index], `canonical_segments[${index}]`);
    } catch (error) {
      throw new SourceIntegrityError(
        "CANONICAL_SEGMENTS_INVALID",
        error instanceof Error ? error.message : `invalid segment ${index}`,
      );
    }
    const start = item.canonical_start;
    const end = item.canonical_end;
    if (!Number.isSafeInteger(start)
      || !Number.isSafeInteger(end)
      || start !== cursor
      || (end as number) < (start as number)
      || (end as number) > canonicalChars) {
      throw new SourceIntegrityError(
        "CANONICAL_SEGMENTS_INVALID",
        `segment ${index} does not continue scalar range at ${cursor}`,
      );
    }
    const originKind = item.origin_kind;
    const originRef = item.origin_ref;
    const transformation = item.transformation;
    if (typeof originKind !== "string" || originKind.length === 0
      || typeof originRef !== "string" || originRef.length === 0
      || typeof transformation !== "string" || transformation.length === 0) {
      throw new SourceIntegrityError(
        "CANONICAL_SEGMENTS_INVALID",
        `segment ${index} has invalid provenance`,
      );
    }
    const rawStart = item.raw_start;
    const rawEnd = item.raw_end;
    if ((rawStart !== undefined || rawEnd !== undefined)
      && (!Number.isSafeInteger(rawStart)
        || !Number.isSafeInteger(rawEnd)
        || (rawStart as number) < 0
        || (rawEnd as number) < (rawStart as number)
        || (rawEnd as number) > rawSize)) {
      throw new SourceIntegrityError(
        "CANONICAL_SEGMENTS_INVALID",
        `segment ${index} has invalid raw range`,
      );
    }
    segments.push({
      canonicalStart: start as number,
      canonicalEnd: end as number,
      originKind,
      originRef,
      transformation,
      ...(rawStart === undefined ? {} : { rawStart: rawStart as number }),
      ...(rawEnd === undefined ? {} : { rawEnd: rawEnd as number }),
    });
    cursor = end as number;
  }
  if (cursor !== canonicalChars) {
    throw new SourceIntegrityError(
      "CANONICAL_SEGMENTS_INVALID",
      `segments stop at ${cursor}, expected ${canonicalChars}`,
    );
  }
  return segments;
}

function validateExcludedRanges(
  input: unknown,
  rawSize: number,
): ExcludedRawRange[] {
  if (!Array.isArray(input)) {
    throw new SourceIntegrityError(
      "EXCLUDED_RANGE_INVALID",
      "excluded_raw_ranges must be an array",
    );
  }
  const ranges: ExcludedRawRange[] = [];
  let previousEnd = 0;
  for (let index = 0; index < input.length; index += 1) {
    const item = record(input[index], `excluded_raw_ranges[${index}]`);
    const policy = item.policy;
    if (typeof policy !== "string"
      || !ALLOWED_EXCLUDED_RAW_POLICIES.has(policy)) {
      throw new SourceIntegrityError(
        "EXCLUDED_POLICY_UNKNOWN",
        `unknown policy at excluded_raw_ranges[${index}]: ${String(policy)}`,
      );
    }
    const start = item.raw_start;
    const end = item.raw_end;
    if (!Number.isSafeInteger(start)
      || !Number.isSafeInteger(end)
      || (start as number) < previousEnd
      || (end as number) <= (start as number)
      || (end as number) > rawSize) {
      throw new SourceIntegrityError(
        "EXCLUDED_RANGE_INVALID",
        `invalid excluded raw range at index ${index}`,
      );
    }
    ranges.push({
      rawStart: start as number,
      rawEnd: end as number,
      policy,
    });
    previousEnd = end as number;
  }
  return ranges;
}

export class SourceLedger implements ScalarSource {
  readonly manifestPath: string;
  readonly rawPath: string;
  readonly canonicalPath: string;
  readonly sourceText: string;
  readonly sourceVersion: string;
  readonly canonicalChars: number;
  readonly canonicalSegments: readonly CanonicalSegment[];
  readonly excludedRawRanges: readonly ExcludedRawRange[];
  readonly coordinates: UnicodeScalarMap;

  private constructor(options: {
    manifestPath: string;
    rawPath: string;
    canonicalPath: string;
    sourceText: string;
    sourceVersion: string;
    canonicalSegments: CanonicalSegment[];
    excludedRawRanges: ExcludedRawRange[];
    coordinates: UnicodeScalarMap;
  }) {
    this.manifestPath = options.manifestPath;
    this.rawPath = options.rawPath;
    this.canonicalPath = options.canonicalPath;
    this.sourceText = options.sourceText;
    this.sourceVersion = options.sourceVersion;
    this.canonicalSegments = options.canonicalSegments;
    this.excludedRawRanges = options.excludedRawRanges;
    this.coordinates = options.coordinates;
    this.canonicalChars = options.coordinates.length;
  }

  static open(manifestPath: string): SourceLedger {
    const resolvedManifestPath = resolve(manifestPath);
    let manifest: Record<string, unknown>;
    try {
      manifest = record(
        JSON.parse(readFileSync(resolvedManifestPath, "utf8")) as unknown,
        "manifest",
      );
    } catch (error) {
      if (error instanceof SourceIntegrityError) {
        throw error;
      }
      throw new SourceIntegrityError(
        "MANIFEST_INVALID",
        error instanceof Error ? error.message : "manifest cannot be parsed",
      );
    }
    if (manifest.schema_version !== SCHEMA_VERSION) {
      throw new SourceIntegrityError(
        "MANIFEST_SCHEMA_UNSUPPORTED",
        `expected ${SCHEMA_VERSION}, got ${String(manifest.schema_version)}`,
      );
    }
    if (manifest.coordinate_unit !== COORDINATE_UNIT) {
      throw new SourceIntegrityError(
        "COORDINATE_UNIT_UNSUPPORTED",
        `expected ${COORDINATE_UNIT}, got ${String(manifest.coordinate_unit)}`,
      );
    }

    const directory = dirname(resolvedManifestPath);
    const rawPath = manifestMemberPath(directory, stringField(manifest, "raw_path"));
    const canonicalPath = manifestMemberPath(
      directory,
      stringField(manifest, "canonical_path"),
    );
    const raw = readFileSync(rawPath);
    const expectedRawSize = integerField(manifest, "raw_size");
    if (raw.length !== expectedRawSize) {
      throw new SourceIntegrityError(
        "RAW_SIZE_MISMATCH",
        `expected ${expectedRawSize} raw bytes, got ${raw.length}`,
      );
    }
    const rawHash = sha256(raw);
    if (rawHash !== stringField(manifest, "raw_sha256")) {
      throw new SourceIntegrityError("RAW_HASH_MISMATCH", "raw SHA-256 does not match");
    }

    const canonical = readFileSync(canonicalPath);
    const canonicalHash = sha256(canonical);
    if (canonicalHash !== stringField(manifest, "canonical_sha256")) {
      throw new SourceIntegrityError(
        "CANONICAL_HASH_MISMATCH",
        "canonical SHA-256 does not match",
      );
    }
    let sourceText: string;
    try {
      sourceText = new TextDecoder("utf-8", { fatal: true }).decode(canonical);
    } catch (error) {
      throw new SourceIntegrityError(
        "CANONICAL_UTF8_INVALID",
        error instanceof Error ? error.message : "canonical source is not UTF-8",
      );
    }
    const coordinates = new UnicodeScalarMap(sourceText);
    const expectedChars = integerField(manifest, "canonical_chars");
    if (coordinates.length !== expectedChars) {
      throw new SourceIntegrityError(
        "CANONICAL_CHAR_COUNT_MISMATCH",
        `expected ${expectedChars} scalars, got ${coordinates.length}`,
      );
    }
    const canonicalSegments = validateSegments(
      manifest.canonical_segments,
      expectedChars,
      raw.length,
    );
    const excludedRawRanges = validateExcludedRanges(
      manifest.excluded_raw_ranges,
      raw.length,
    );

    return new SourceLedger({
      manifestPath: resolvedManifestPath,
      rawPath,
      canonicalPath,
      sourceText,
      sourceVersion: canonicalHash,
      canonicalSegments,
      excludedRawRanges,
      coordinates,
    });
  }

  slice(start: number, end: number): string {
    return this.coordinates.slice(start, end);
  }
}
